### ISSUE TITLE
Non-Square images (Aspect ratio 16:9) for Training the YoloV8 network

### ISSUE THREAD

**Issue by vanguard478** [ultralytics/ultralytics#3815]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hi, 
I am trying to train YoloV8 on my custom dataset, in which the image sizes are 1920 by 1080 i.e 16:9 aspect ratio. Also, the same resolution will be used for inference. I cannot possibly convert them to a square aspect ratio without distorting features in the image.  
I found this [issue](https://github.com/ultralytics/yolov5/issues/11768#issuecomment-1609404382) with training on non-square image size in which it was mentioned to use give the image dimension as `imgsz` and `rect=True`. 
When I am using this command with v8 
```python
from ultralytics import YOLO

# Load a model
yolo_model = YOLO('yolov8n.yaml')
yolo_model = YOLO('yolov8n.pt')  # load a pretrained model 
yolo_model = YOLO('yolov8n.yaml').load('yolov8n.pt')  
data_yaml = 'weld_spatter_data.yaml'

# Train the model
yolo_model.train(data=data_yaml, epochs=100, imgsz=[1920, 1080],rect=True)
```
I am getting this Warning:
`WARNING ⚠️ updating to 'imgsz=1920'. 'train' and 'val' imgsz must be an integer, while 'predict' and 'export' imgsz may be a [h, w] list or an integer, i.e. 'yolo export imgsz=640,480' or 'yolo export imgsz=640'`

How can I ensure the best results, if the images are non-square for both training and inference? I can think of the following options:
1. Split the image into two vertical images (960px by 1080px each) and resize it to square images? This will have slightly less distortion when converted to a square aspect ratio ( 1 - 960px/1080px = 0.11 )  compared to resizing the original images ( 1 - 1920px/1080px = -0.77) . 
2. Can I add padding instead, to make it of square resolution and if yes what background (padding) colour should I choose for  YoloV8?
3. In general what is the preferred method for yoloV8 out of option 1 and option 2 ? Padding or cropping?

**Comment by xaristeidou** [comment]
YOLO doesn't need to convert to square the image. It work for different aspect ratios and doesn't have a problem. Why would you like to use square images there is no need for that. 

Also I think YOLOv8 can train images up to imgsz = 1280, I think can accept arguments more than that, but not totally sure.

**Comment by glenn-jocher** [comment]
@vanguard478 you're correct in noting that YOLOv8 doesn't need square images for training or inference. YOLOv8 is designed to handle different aspect ratios, and therefore there's no obligation to convert your images to a square aspect ratio. Utilizing your images in their original aspect ratio of 16:9 can work without issue.

The `imgsz` parameter indicates the input dimension for your training images which can be specified as a single integer for square images or as a list `[h, w]` for rectangular images. The `--rect` flag is typically used in tandem with `imgsz` and helps to preserve the original aspect ratios of your images during training.  This can be particularly beneficial when the aspect ratios in your validation set correspond to those in your training set.

In terms of image size, YOLOv8 does support training images with sizes larger than 1280. However, you will need to ensure your hardware can handle the increased computational requirement.

Regarding your proposed strategies:
1. Splitting the image into two vertical images: This might introduce complications related to maintaining object continuity and context. Also, information may be lost at the borders where the split occurs.
2. Adding padding: This is a viable option, which eliminates the need for image cropping and thus preserves the entirety of your content. Your padding color should ideally be a value that doesn't exist in your dataset, to help avoid confusion.

In general, neither method is strictly "preferred" - it really depends on your specific use-case and dataset. It's always a good idea to try out both methods and evaluate their performance on your validation set. Always consider what each approach means for your particular task when choosing the best strategy.

**Comment by vanguard478** [comment]
@glenn-jocher  Thank a lot you for the reply. I understand that YOLO can handle different image resolutions and I was able to do the inference on different resolutions. I am concerned about the **initial image resolution** it is taking for **training the network.**
I tried inputting the `imgsz `as a list [1920,1280] but I was shown the warning during the start of the training as shown below: 
`WARNING ⚠️ updating to 'imgsz=1920'. 'train' and 'val' imgsz must be an integer, while 'predict' and 'export' imgsz may be a [h, w] list or an integer, i.e. 'yolo export imgsz=640,480' or 'yolo export imgsz=640' `  

- So my question is are the images being resized during the training to 1920px by 1920py into a square or is it maintaining the original aspect ratio of 16:9  i.e. keeping the images at 1920px by 1280px. **What exactly is this warning telling me?**
 
- If the images are resized automatically, **how can I ensure that the aspect ratio isn't changed during training** and the network is trained on the original image resolution (and so the aspect ratio)? I have tried this command to training but I don't think it is working correctly as it is giving me the warning as above:
`yolo_model.train(data=data_yaml, epochs=100, imgsz=[1920, 1080],rect=True)`

> The `imgsz` parameter indicates the input dimension for your training images which can be specified as a single integer for square images or as a list `[h, w]` for rectangular images. The `--rect` flag is typically used in tandem with `imgsz` and helps to preserve the original aspect ratios of your images during training. This can be particularly beneficial when the aspect ratios in your validation set correspond to those in your training set.

**Comment by NMVRodrigues** [comment]
@vanguard478 Hello, I'm facing the exact same problem and going through the Docs did not provide any clarification. Did you end up finding out if the aspect ratio is kept or if it is just "squaring" the image?

**Comment by vanguard478** [comment]
> @vanguard478 Hello, I'm facing the exact same problem and going through the Docs did not provide any clarification. Did you end up finding out if the aspect ratio is kept or if it is just "squaring" the image?

The YOLOv8 doesn't change the aspect ratio when you specify the `imgsz=640` parameter , it will maintain the aspect ratio by making the larger dimension of the image equal to 640px and the smaller dimension will be padded.

**Comment by NMVRodrigues** [comment]
@vanguard478 Thanks!

**Comment by glenn-jocher** [comment]
@NMVRodrigues you're welcome! If you encounter any issues or have more questions as you proceed, don't hesitate to reach out. Happy training with YOLOv8!

**Comment by Buckler89** [comment]
Hello,

I've been exploring the data augmentation capabilities for classification tasks within the codebase and noticed that the current augmentation strategy seems to be configured to only produce square images. This observation is based on the implementation found here: [augment.py#L1096](https://github.com/ultralytics/ultralytics/blob/61088b43d763a91c614134f295cf762f5beaba76/ultralytics/data/augment.py#L1096).

It appears that the size parameter for classification does not accept a list, implying that only square image dimensions are supported. This limitation raises a couple of questions:

1. Is there a specific reason behind the design choice to restrict image dimensions to square shapes for classification tasks?
2. Would it be possible to modify the augmentation strategy to accommodate non-square image dimensions, and if so, are there any planned updates in this regard?

Understanding the reasoning behind this design choice would be incredibly helpful, as would any guidance on how to potentially adjust the code to support non-square images for classification purposes.

Thank you for your time and assistance.

**Comment by lorenzoriano** [comment]
> Hello,
> 
> I've been delving into the data loading and augmentation functionalities tailored for classification tasks within the repository and encountered a couple of issues related to handling non-square image dimensions that I hope to get some clarity on.
> 
> 1. **Augmentation Strategy Limitation:** Based on the code found at [augment.py#L1096](https://github.com/ultralytics/ultralytics/blob/61088b43d763a91c614134f295cf762f5beaba76/ultralytics/data/augment.py#L1096), it seems the augmentation strategy for classification tasks is designed to only produce square images, as the size parameter doesn't accept a list for varying dimensions. Is there a particular reason for this limitation? Moreover, is there a workaround or a future plan to support non-square dimensions in classification augmentation?
> 2. **System Warning on Image Size Specification:** Following an attempt to use non-square images for cropping, I encountered a system warning that suggests a fundamental constraint:
> 
> ```
> WARNING ⚠️ updating to 'imgsz=640'. 'train' and 'val' imgsz must be an integer, while 'predict' and 'export' imgsz may be a [h, w] list or an integer, i.e. 'yolo export imgsz=640,480' or 'yolo export imgsz=640'
> ```
> 
> As highlighted by @vanguard478, this warning indicates that the 'train' and 'val' phases require an integer for `imgsz`, thereby preventing the propagation of a non-square shape to the augmentation strategy function.
> 
> These points suggest a systemic limitation for utilizing non-square images in classification tasks, impacting the flexibility of data augmentation and potentially the effectiveness of the model training process.
> 
> * Could you provide insight into the rationale behind these constraints?
> * Are there any recommendations for handling non-square images within the current framework, or are there plans to enhance the system's capability in this respect?
> 
> Thank you for addressing these concerns. Your guidance would greatly help in optimizing our approach to handling diverse image dimensions in classification models.

Could @glenn-jocher comment on this?

**Comment by glenn-jocher** [comment]
@lorenzoriano hello,

Thanks for reaching out and bringing up your observations on handling non-square images within YOLOv8. Here's a bit of insight:

1. **Augmentation Strategy**: The current augmentation strategy primarily producing square images was initially chosen for simplicity and consistency across different tasks. It's less about constraints and more about providing a stable baseline for most use cases. However, we understand the necessity for flexibility in real-world applications. While there isn't a built-in workaround just yet, we're always evaluating how to improve and potentially include support for non-square dimensions in future updates.

2. **Image Size Specification**: The warning you're seeing essentially serves as a guideline to ensure consistency during training and validation phases. For `predict` and `export`, the flexibility is greater due to varied real-world deployment scenarios. 

For non-square images in the current framework, one approach is to pre-process your images with padding to maintain aspect ratio, or potentially resize them in a way that least impacts your specific use-case. These are not perfect solutions, but they can be a stopgap until more flexible options are available.

We're always looking to enhance YOLOv8's capabilities based on community feedback. Your points are valid, and addressing them helps us all in pushing the boundaries of what's possible with the framework. Stay tuned for updates, and thanks again for your in-depth engagement!

P.S. For direct code change suggestions or feature requests, don't hesitate to contribute through GitHub issues or pull requests. Your contributions can significantly shape the evolution of YOLOv8. 🚀

**Comment by fromm1990** [comment]
I may be totally wrong, but maybe someone could prove or disprove my observation.

The [classify_transforms](https://github.com/ultralytics/ultralytics/blob/42416bc608b49cbdcb091e81adbb4caaf11eac38/ultralytics/data/augment.py#L1137C5-L1137C24) function uses the pytorch [Resize](https://pytorch.org/vision/main/generated/torchvision.transforms.Resize.html) transform in combination with the [CenterCrop](https://pytorch.org/vision/0.15/generated/torchvision.transforms.CenterCrop.html#torchvision.transforms.CenterCrop) transform.

Let's assume we have a dataset where the images are of various aspect ratios. We therefore provide the a single scalar value as the image size  e.g. 680 (the same should be true if a quadratic tuple, i.e. `(680, 680)` is provided.

The following is documented for the size parameter of the Resize transform function: 
> Desired output size. If size is a sequence like (h, w), output size will be matched to this. If size is an int, smaller edge of the image will be matched to this number. i.e, if height > width, then image will be rescaled to (size * height / width, size).

This should mean that if we have an image (380px, 240px) it will get resized to (1077px, 680px). The resized image will then be ran through the CenterCrop transform function. This will result in a cropped image (680px, 680px) where 37% of the original information is lost.

Given that my understanding of the code is correct, is there a reason why images are not simply scaled down to make the longest edge fit the size and then pad the rest to reach the 680 x 680?

**Comment by glenn-jocher** [comment]
Hey there!

You've got a keen eye for detail! 🕵️ Your interpretation of the [`classify_transforms`](https://github.com/ultralytics/ultralytics/blob/42416bc608b49cbdcb091e81adbb4caaf11eac38/ultralytics/data/augment.py#L1137C5-L1137C24) function is on point. When images with varying aspect ratios are resized to a single dimension (like 680), and then center-cropped, a portion of the original image information can indeed be lost. This process ensures a consistent input size for models at the cost of some image content.

The choice to resize and center-crop rather than scale down and pad relates to maintaining the integrity of the visual information. Scaling down and padding could introduce noise or irrelevant content (padding) that might affect the model's ability to learn the important features.

That said, your suggestion about scaling to the longest edge and then padding has its merits, especially in preserving the full context of the image. The decision between both approaches can depend on the specific dataset and task requirements. 

For those who'd prefer scaling and padding, this could be implemented custom in their preprocessing workflow, though it's not the current default approach in our pipeline. We're always open to improvements, so we'll consider this feedback for future versions!

Thanks for bringing this up, and keep those observations coming! 🚀

**Comment by roy-orfaig** [comment]
Can I train in yolo8 custom dataset the include different size of images? 2048X2046 and 1536X2048?

**Comment by glenn-jocher** [comment]
Yes, YOLOv8 handles varying image resolutions by resizing and padding while preserving aspect ratios during training.

**Comment by roy-orfaig** [comment]
Thanks! But is it recommended to separate different models by image size? Does training the model with various image sizes negatively affect metric performance?

**Comment by glenn-jocher** [comment]
Thank you for your contribution to BOT-SORT! 🎉 For YOLOv8, we recommend using a **consistent** `imgsz` during training as the model automatically resizes images while preserving aspect ratios with padding. Mixed training sizes can introduce inconsistency in feature learning, potentially impacting metrics. For optimal results, maintain a fixed `imgsz` aligned with your deployment resolution. Learn more in our [Training Documentation](https://docs.ultralytics.com/modes/train/).
